### ISSUE TITLE
Functionality for calculating mean and standard deviation for image datasets

### ISSUE THREAD

**Issue by dhananjaisharma10** [anjandeepsahni/torchutils#1]
The functionality for the above feature should be added. I was hoping to contribute it in a way which is generic to most problems. The challenge that I'm facing is to incorporate it with a dataset class such that the mean and std are calculated before even calling `__getitem__`.

**Comment by anjandeepsahni** [comment]
We can define a function that accepts the `dataloader` object. It will then iterate over all items in the dataset using `__getitem__` and calculate the mean. The user can specify a parameter in their dataset class, say `self.transforms`, to control whether to return raw image or transformed image from `__getitem__`.

Initially `self.transforms` will be `None` and `__getitem__` will return raw image. The user can then call the mean and std calculation API to get the data statistics. Subsequently `self.transforms` in dataset class can be updated to incorporate this mean and std to normalize the data.

Flow would be something like this:


- Create dataset object with dataset.transforms as None.
-- This means __getitem__ returns raw data.
- Create dataloader.
- Call API and pass dataloader as parameter.
- API will loop over all data and calculate mean and std.
-- API will return the mean and std (for each channel).
- Update dataset.transforms to use torchvision.transforms.Normalize with calculated mean and std.
- User dataloader for training, but now __getitem__ returns transformed data.


It is similar to how you would control/prevent the application of some transformations (data augmentation) during testing time.

**Comment by dhananjaisharma10** [comment]
I was planning to do something similar but this technique has its own caveats, the biggest one being:
- Both mean and std will be calculated every single time the model is trained, thereby slowing down the process.
I feel we should cut the redundant computation by storing the values in a file and simply checking for  it at every training or evaluation call.

**Comment by anjandeepsahni** [comment]
That seems like something user should manage. But we can make the API more useful with the following set of parameters:

```
def get_dataset_mean_std(dataloader=None, dump_result=False, dump_file_path=None, load_from_file=False):
    if load_from_file:
        # load mean and std from dump_file_path
    else:
        # calculate mean and std
        if dump_result:
            # dump to dump_file_path
```

Obviously we will need to check that the parameter combination that the user has passed is valid.

**Comment by anjandeepsahni** [comment]
Support has been added to master branch. Will be part of next release.
